### PR TITLE
[DNM] compiler: add crate compiler, compiles client side code

### PR DESCRIPTION
**Pull request by overvenus**
PTAL @BusyJay @siddontang @disksing @andelf 

Please review files under path `compiler`.

### CODE REVIEW

**Comment by disksing** [comment]
LGTM.

File: compiler/src/util.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/// Adjust method name to follow the rust's style.
+pub fn snake_name(name: &str) -> String {
```
Comment (andelf) [inline compiler/src/util.rs]: wired function name. do not invent terms. 
use to_snake_case or something.

File: src/server.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        self.slots_per_cq = slots;
+        self
+    }
+
+    pub fn build(mut self) -> Server {
+        let args = self.args.map_or_else(ptr::null, |args| args.as_ptr());
+        unsafe {
+            let server = grpc_sys::grpc_server_create(args, ptr::null_mut());
+            let bind_addrs: Vec<_> = self.addrs.drain(..).map(|(host, port)| {
+                let addr = format!("{}:{}\0", host, port);
```
Comment (andelf) [inline src/server.rs]: CString

File: grpc-sys/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    Shutdown,
+}
+
+#[repr(C)]
+pub enum GrpcCompressionLevel {
+    None = 0,
+    Low,
+    Med,
+    High,
+    Count
```
Comment (andelf) [inline grpc-sys/src/lib.rs]: you've rewrite whole thing? 
or you've done a force merge?

**Comment by disksing** [comment]
I'm going to open another PR contains only compiler code.

**Pull request closed without merging**